unblob/handlers/filesystem/yaffs/utils.py

```diff
@@ -0,0 +1,574 @@
+        # checksum bytes.
+        #
+        # Thus, the signature for identifying the next page section (and hence, the end of the
+        # spare data section) becomes: [the 4 bytes starting at offset 0x804] + 0xFFFF
+        #
+        # Note that this requires at least one non-empty subdirectory; in practice, any Linux
+        # file system should meet this requirement, but one could create a file system that
+        # does not meet this requirement.
+
+        object_id_offset = 4
+        object_id_start = page_size + ecc_offset + object_id_offset
+        object_id_end = object_id_start + 4
+        spare_signature = self.file[object_id_start:object_id_end] + b"\xFF\xFF"
+
+        config.spare_size = (
+            self.file[object_id_end : object_id_end + page_size].find(spare_signature)
+            + object_id_offset
+            + ecc_offset
+        )
+
+        # Sanity check the spare size, make sure it looks legit
+        if config.spare_size not in VALID_SPARE_SIZES:
+            raise InvalidInputFormat(
+                "Auto-detection failed: Detected an unlikely spare size: %d"
+                % config.spare_size
+            )
+
+        return config
+
+    def insert_entry(self, entry: YAFFSEntry):
+        try:
+            self.file_entries.create_node(
+                entry.object_id,
+                entry.object_id,
+                data=entry,
+                parent=entry.parent_obj_id,
+            )
+        except NodeIDAbsentError:
+            pass
+        except DuplicatedNodeIdError:
+            pass
+
+    def get_entry(self, object_id: int):
+        try:
+            entry = self.file_entries.get_node(object_id)
+            if entry:
+                return entry.data
+        except NodeIDAbsentError:
+            pass
+        return None
+
+    def resolve_path(self, entry: YAFFSEntry) -> Path:
+        resolved_path = Path(entry.name)
+        parent = self.file_entries[entry.parent_obj_id].data
+        if parent is not None:
+            return self.resolve_path(parent).joinpath(resolved_path)
+        return resolved_path
+
+    def get_file_bytes(self, entry: YAFFSEntry) -> Iterable[bytes]:
+        for chunk in entry.get_chunks():
+            self.file.seek(
+                entry.start_offset
+                + (
+                    (chunk.chunk_id - 1)
+                    * (self.config.page_size + self.config.spare_size)
+                ),
+                io.SEEK_SET,
+            )
+            yield self.file.read(chunk.byte_count)
+
+    def extract(self, outdir: Path):
+        self.file_entries.show()
+        for entry in [
+            self.file_entries.get_node(node).data
+            for node in self.file_entries.expand_tree(mode=Tree.DEPTH)
+        ]:
+            if entry is None:
+                continue
+            self.extract_entry(entry, outdir)
+
+    def extract_entry(self, entry: YAFFS2Entry, outdir: Path):  # noqa: C901
```

Comment:
is entry always YAFFS2Entry, what if we extract v1?

Reply:
YAFFSv1 does not support special files and if I set this type to `YAFFSEntry` pyright complaints that special files fields (e.g. `yst_mode`) does not exist.

Could be solved by a better class design for entries.